app/assets/javascripts/discourse/app/lib/hashtag-autocomplete.js

```diff
@@ -42,17 +84,80 @@ function _setup($textArea, siteSettings, afterComplete) {
     template: findRawTemplate("category-tag-autocomplete"),
     key: "#",
     afterComplete,
-    transformComplete: (obj) => {
-      return obj.text;
-    },
+    transformComplete: (obj) => obj.text,
     dataSource: (term) => {
       if (term.match(/\s/)) {
         return null;
       }
       return searchCategoryTag(term, siteSettings);
     },
-    triggerRule: (textarea, opts) => {
-      return categoryHashtagTriggerRule(textarea, opts);
-    },
+    triggerRule: (textarea, opts) => hashtagTriggerRule(textarea, opts),
   });
 }
+
+let searchCache = {};
+let searchCacheTime;
+let currentSearch;
+
+function _updateSearchCache(term, results) {
+  searchCache[term] = results;
+  searchCacheTime = new Date();
+  return results;
+}
+
+function _searchGeneric(term, siteSettings, context) {
+  if (currentSearch) {
+    currentSearch.abort();
+    currentSearch = null;
+  }
+  if (new Date() - searchCacheTime > 30000) {
+    searchCache = {};
+  }
+  const cached = searchCache[term];
+  if (cached) {
+    return cached;
+  }
+
+  return new Promise((resolve) => {
+    let timeoutPromise = isTesting()
+      ? null
+      : discourseLater(() => {
```

Comment:
discourseLater is already including the isTesting check now

Reply:
Cool I will fix that up thanks for letting me know 👍

Reply:
I actually had to change this back, the promise fires too fast and the autocomplete never shows if we use `discourseLater` as is, which makes it impossible to test. We can refactor this later.